src/models/room.ts

```diff
@@ -904,6 +929,71 @@ export class Room extends TypedEventEmitter<EmittedEvents, RoomEventHandlerMap>
+     * historical messages are imported into the room via MSC2716 `/batch_send
+     * because the client may already have that section of the timeline loaded.
+     * We need to force the client to throw away their current timeline so that
+     * when they back paginate over the area again with the historical messages
+     * in between, it grabs the newly imported messages. We can listen for
+     * `UNSTABLE_MSC2716_MARKER`, in order to tell when historical messages are ready
+     * to be discovered in the room and the timeline needs a refresh. The SDK
+     * emits a `RoomEvent.historyImportedWithinTimeline` event when we detect a
+     * valid marker and can check the needs refresh status via
+     * `room.getTimelineNeedsRefresh()`.
+     */
+    public async refreshLiveTimeline(): Promise<void> {
+        const liveTimelineBefore = this.getLiveTimeline();
+        const forwardPaginationToken = liveTimelineBefore.getPaginationToken(EventTimeline.FORWARDS);
+        const eventsBefore = liveTimelineBefore.getEvents();
+        const mostRecentEventInTimeline = eventsBefore[eventsBefore.length - 1];
+        logger.log(
+            `[refreshLiveTimeline] for room ${this.roomId} at ` +
+            `mostRecentEventInTimeline=${mostRecentEventInTimeline.getId()} ` +
+            `forwardPaginationToken+${forwardPaginationToken}`,
+        );
+
+        // Empty out all of `this.timelineSets` but still keeps the same
+        // `timelineSet` references around so the React code updates properly
+        // and doesn't ignore the room events we emit because it checks that the
+        // `timelineSet` references are the same. We need the `timelineSet`
+        // empty so that the `client.getEventTimeline(...)` call later, will
+        // call `/context` and create a new timeline instead of returning the
+        // same one.
+        this.resetLiveTimeline(null, null);
+
+        // Get a reference to the emptied out `timelineSet`
+        //
+        // TODO: Do we want to use `this.getLiveTimeline().getTimelineSet()`
+        // instead? I think it's the same thing but what's more right?
+        const timelineSet = this.getUnfilteredTimelineSet();
+
+        // Use `client.getEventTimeline(...)` to construct a new timeline from a
+        // `/context` response state and events for the most recent event before
+        // we reset everything. The `timelineSet` we pass in needs to be empty
+        // in order for this function to call `/context` and generate a new
+        // timeline.
+        const newTimeline = await this.client.getEventTimeline(timelineSet, mostRecentEventInTimeline.getId());
+        // Set the pagination token back to the live sync token instead of using
+        // the `/context` historical token so that it matches the next response
+        // from `/sync` and we can properly continue the timeline.
+        newTimeline.setPaginationToken(forwardPaginationToken, EventTimeline.FORWARDS);
```

Comment:
This seems like it could be problematic, by mixing pagination tokens between timelines you're basically acknowledging that events may be missing or duplicated due to race issues between /sync and /context

Reply:
We could add some better merging logic here. If after `getEventTimeline` runs, we see another timeline (indicating that a sync happened in between), we can skip setting the forward pagination token and setting it as the live timeline.

We can rely on the latest sync and since we cleared out the other timelines, it will paginate back from the latest sync.

Definitely want to add some tests for these scenarios, https://github.com/matrix-org/matrix-react-sdk/pull/8354#discussion_r852588155

Reply:
In https://github.com/matrix-org/matrix-react-sdk/pull/8354, we have an e2e Cypress test for a `/sync` racing between the `this.resetLiveTimeline(null, null);` and when the `client.getEventTimeline()` `/context` request completes. Everything seems to behave as expected ✅

The live timeline has a forward pagination token of `null` so it merges perfectly without issue.

Here is a breakdown of the timelines:
```
// After `this.resetLiveTimeline(null, null);`
[
    {
        "0-!OIGyXdhYVmWEAByyII:localhost:2022-05-26T02:39:10.737Z--liveTimeline--f=null--b=null": []
    }
]

// After racey `/sync` completes
[
    {
        "0-!OIGyXdhYVmWEAByyII:localhost:2022-05-26T02:39:10.737Z--liveTimeline--f=null--b=null": [
            "$LpjtBslib3J0kPzL1J8EbiAg7mNCNQfHoki7nh9ejH8"
        ]
    }
]

// After we use `client.getEventTimeline(...)` to construct a new timeline from a `/context` response
[
    {
        "0-!OIGyXdhYVmWEAByyII:localhost:2022-05-26T02:39:10.737Z--liveTimeline--f=null--b=t10--6_0_0_0_0_0_0_0_0": [
            "$8-8QjB-U0BRFcmpY39FkMpdt4-R8ux5XbHL0B0LXZ00",
            "$7DrVapInIVSKaM8MWQZDi15_ruTIUQfyY4grQu-NOiw",
            "$bFyHxdn9Y8wosMxnvA-HdXaGSHpT1VridJnGUvsBGWU",
            "$wvoRm9JI5uMeWxVm49mny2YtthfYKdEAZ9g3-gbeCJc",
            "$kOFxIReYWJK-M7JmNyFGINkq4Y7FBlD3jMEFzk1GqNg",
            "$3ZcZZe9MKr35wVVkI-1HoGem-WRdyEtOMJttCQq9QUU",
            "$LpjtBslib3J0kPzL1J8EbiAg7mNCNQfHoki7nh9ejH8"
        ]
    }
]

// After things settle
[
    {
        "0-!OIGyXdhYVmWEAByyII:localhost:2022-05-26T02:39:10.737Z--liveTimeline--f=null--b=t10--6_0_0_0_0_0_0_0_0": [
            "$8-8QjB-U0BRFcmpY39FkMpdt4-R8ux5XbHL0B0LXZ00",
            "$7DrVapInIVSKaM8MWQZDi15_ruTIUQfyY4grQu-NOiw",
            "$bFyHxdn9Y8wosMxnvA-HdXaGSHpT1VridJnGUvsBGWU",
            "$wvoRm9JI5uMeWxVm49mny2YtthfYKdEAZ9g3-gbeCJc",
            "$kOFxIReYWJK-M7JmNyFGINkq4Y7FBlD3jMEFzk1GqNg",
            "$3ZcZZe9MKr35wVVkI-1HoGem-WRdyEtOMJttCQq9QUU",
            "$LpjtBslib3J0kPzL1J8EbiAg7mNCNQfHoki7nh9ejH8"
        ]
    }
]
```

Reply:
Let me know if there are other scenarios you are particularly worried about and we can add an e2e test for them.

Reply:
It'd be better to have a test at the js-sdk layer for this, given that a change in the js-sdk being merged should not cause cascading failures to downstream consumers

Reply:
👍 Added tests in the `matrix-js-sdk` and fixed up the merging scenario which is tested harder, https://github.com/matrix-org/matrix-js-sdk/pull/2299#discussion_r886552876